### ISSUE TITLE
Navigator.canPop(context) return true on initial route page after changing screen orientation

### ISSUE THREAD

**Issue by mozochek** [flutter/flutter#66531]
## Code where bug occur
```dart
// This code in home screen, which is initial route of app
if (Navigator.canPop(context)) {
  actions.add(
    HomeIconButton(),
  );
}
```
## Steps to Reproduce

1. Open initial route screen
2. Move to next screen
3. Change screen orientation
4. Return back

**Expected results:** Navigator.canPop() return false and HomeIconButton does not appear on app bar

**Actual results:** Navigator.canPop() return true and HomeIconButton appear on app bar

## Possible solution
The [AppBar](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/app_bar.dart) class contains this code:

```dart
final ModalRoute<dynamic> parentRoute = ModalRoute.of(context);
...
bool canPop = parentRoute?.canPop ?? false;
...
if (!hasEndDrawer && canPop)
          leading = useCloseButton ? const CloseButton() : const BackButton();
```
And this code works well - BackButton does not appear on app bar after changing screen orientatoion.

The [Navigator](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/navigator.dart).canPop(context) current realisation is: 
```dart
static bool canPop(BuildContext context) {
  final NavigatorState? navigator = Navigator.of(context, nullOk: true);
  return navigator != null && navigator.canPop();
}
```
I think the AppBar's code need integrate to Navigator.canPop(context)

**Comment by TahaTesser** [comment]
Hi @mozochek 
Can you please provide your `flutter doctor -v` your `flutter run --verbose` and a complete reproducible minimal code sample
Thank you

**Comment by mozochek** [comment]
Hello @TahaTesser 

# Logs
<details>
<summary>flutter doctor -v</summary>

[✓] Flutter (Channel stable, 1.20.4, on Microsoft Windows [Version 6.1.7601], locale ru-RU)
    
    • Flutter version 1.20.4 at C:\Important\flutter
    • Framework revision fba99f6cf9 (10 days ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)

    • Android SDK at C:\Users\mozochek\AppData\Local\Android\Sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = C:\Users\mozochek\AppData\Local\Android\Sdk
    • Java binary at: C:\Important\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[✓] Android Studio (version 3.6)

    • Android Studio at C:\Important\Android Studio
    • Flutter plugin version 49.0.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[✓] Connected device (1 available)

    • COL L29 (mobile) • CLCDU18731001969 • android-arm64 • Android 10 (API 29)

• No issues found!
</details>

[flutter run --verbose](https://github.com/flutter/flutter/files/5281127/flutter.run.--verbose.txt)

# Steps to produce
1. Run app with code in the sample
2. Press button "second page" on home screen
3. Rotate screen to the oppoiste (lanscape or portrait)
4. Press back arrow icon button on app bar or button "rotate screen and go back" on screen

**Expected results:** Home icon in right app bar side does not apper after reproducing steps above

**Actual results:** Home icon appear in right app bars side

# Code

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(Application());
}

class Application extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: HomeScreen(),
    );
  }
}

class CustomAppBar extends PreferredSize {
  final String title;

  CustomAppBar({@required this.title});

  @override
  Size get preferredSize => Size.fromHeight(AppBar().preferredSize.height);

  @override
  Widget build(BuildContext context) {
    final List<Widget> actions = [];
    final bool isMobile = MediaQuery.of(context).size.shortestSide < 600.0; // without this everything works well

    if (Navigator.canPop(context)) {
      actions.add(
        IconButton(
          icon: Icon(Icons.home),
          onPressed: () {
            Navigator.pushAndRemoveUntil(
              context,
              MaterialPageRoute(
                builder: (context) => HomeScreen(),
              ),
              (route) => false,
            );
          },
        ),
      );
    }

    return AppBar(
      title: Text(
        title,
      ),
      actions: actions,
      centerTitle: true,
    );
  }
}

class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: CustomAppBar(
        title: 'home screen',
      ),
      body: Center(
        child: RaisedButton(
          onPressed: () => {
            Navigator.push(
              context,
              MaterialPageRoute(
                builder: (context) => SecondScreen(),
              ),
            ),
          },
          child: Text('second page'),
        ),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: CustomAppBar(title: 'second screen'),
      body: Center(
        child: RaisedButton(
          onPressed: () => {
            Navigator.pop(context),
          },
          child: Text('rotate screen and go back'),
        ),
      ),
    );
  }
}

```

</details>

**Comment by TahaTesser** [comment]
Hi @mozochek 
I just tried to reproduce the issue based on your steps, I don't see any home icon when pressing the back button
![ezgif com-gif-maker](https://user-images.githubusercontent.com/48603081/94408470-7b467480-0192-11eb-8218-5dbc92ccb3fa.gif)


<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.20.4, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.20.4 at /Users/tahatesser/Code/flutter_stable
    • Framework revision fba99f6cf9 (13 days ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

 
[✓] Connected device (2 available)            
    • sdk gphone x86 arm (mobile) • emulator-5554             • android-x86 •
      Android 11 (API 30) (emulator)
    • Taha’s iPhone (mobile)      • 00008020-001059882212002E • ios         •
      iOS 14.0.1

• No issues found!
```

</details>

**Comment by mozochek** [comment]
> Hi @mozochek
> I just tried to reproduce the issue based on your steps, I don't see any home icon when pressing the back button
> ![ezgif com-gif-maker](https://user-images.githubusercontent.com/48603081/94408470-7b467480-0192-11eb-8218-5dbc92ccb3fa.gif)
> 
> flutter doctor -v

Hi @TahaTesser 
Maybe I incorrectly describe 3rd step. Instead of only rotating screen you need apply rotation, so application should rebuild for new orientation. You just rotate emulator screen, but you don't rotate emulator GUI (for applying GUI rotation in emulator you need to press little button with 2 arrows which appears after emulator screen rotation)
![ezgif-7-58e889a4c652](https://user-images.githubusercontent.com/50376680/94517282-4a1c8180-0230-11eb-96d9-3232c65c3778.gif)

**Comment by TahaTesser** [comment]
Hi @mozochek 
I can reproduce and removing this line does solve the issue, which is bizarre, since this is totally an used  line of code
```dart
    final List<Widget> actions = [];
    final bool isMobile = MediaQuery.of(context).size.shortestSide <
        600.0; // without this line everything works well

    if (Navigator.canPop(context)) {
      actions.add(
        IconButton(
          icon: Icon(Icons.home),
          onPressed: () {
            Navigator.pushAndRemoveUntil(
              context,
              MaterialPageRoute(
                builder: (context) => HomeScreen(),
              ),
              (route) => false,
            );
          },
        ),
      );
```

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(Application());
}

class Application extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: HomeScreen(),
    );
  }
}

class CustomAppBar extends PreferredSize {
  final String title;

  CustomAppBar({@required this.title});

  @override
  Size get preferredSize => Size.fromHeight(AppBar().preferredSize.height);

  @override
  Widget build(BuildContext context) {
    final List<Widget> actions = [];
    final bool isMobile = MediaQuery.of(context).size.shortestSide <
        600.0; // without this line everything works well

    if (Navigator.canPop(context)) {
      actions.add(
        IconButton(
          icon: Icon(Icons.home),
          onPressed: () {
            Navigator.pushAndRemoveUntil(
              context,
              MaterialPageRoute(
                builder: (context) => HomeScreen(),
              ),
              (route) => false,
            );
          },
        ),
      );
    }

    return AppBar(
      title: Text(
        title,
      ),
      actions: actions,
      centerTitle: true,
    );
  }
}

class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: CustomAppBar(
        title: 'home screen',
      ),
      body: Center(
        child: RaisedButton(
          onPressed: () => {
            Navigator.push(
              context,
              MaterialPageRoute(
                builder: (context) => SecondScreen(),
              ),
            ),
          },
          child: Text('second page'),
        ),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: CustomAppBar(title: 'second screen'),
      body: Center(
        child: RaisedButton(
          onPressed: () => {
            Navigator.pop(context),
          },
          child: Text('rotate screen and go back'),
        ),
      ),
    );
  }
}

```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.20.4, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.20.4 at /Users/tahatesser/Code/flutter_stable
    • Framework revision fba99f6cf9 (2 weeks ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

 
[✓] Connected device (1 available)            
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86 • Android 11 (API 30) (emulator)

• No issues found!
```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.22.0-10.0.pre.405, on Mac OS X 10.15.7 19H2
    x86_64, locale en-GB)
    • Flutter version 1.22.0-10.0.pre.405 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 9d7bdb5903 (4 hours ago), 2020-09-28 22:46:21 -0700
    • Engine revision 438685dad0
    • Dart version 2.11.0 (build 2.11.0-174.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

 
[✓] Connected device (4 available)            
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86    • Android 11
      (API 30) (emulator)
    • macOS (desktop)             • macos         • darwin-x64     • Mac OS X
      10.15.7 19H2 x86_64
    • Web Server (web)            • web-server    • web-javascript • Flutter
      Tools
    • Chrome (web)                • chrome        • web-javascript • Google
      Chrome 85.0.4183.121

• No issues found!
```

</details>

**Comment by raulmabe** [comment]
I am having a similar issue, but only happens when a `bloc` from `flutter_bloc` is rebuilt... I will try to provide a minimal reproducible code whenever I can.

**Comment by tithamane** [comment]
In my case, it seems to happen on hot-reload. I've added the code below, just create new project and replace the code in "main.dart" with it.

Create the app and run it with the code below. If you do not update the colour on the second page then canPop on the first page will continue to return false but if you did update the second page to trigger hot-reload then on the first page the "canPop" will start to return true.


`import 'package:flutter/material.dart';

const Color darkBlue = Color.fromARGB(255, 18, 32, 47);

void main() {
  runApp(const HotReloadProblemApp());
}

class HotReloadProblemApp extends StatelessWidget {
  const HotReloadProblemApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: FirstScreen());
  }
}

class FirstScreen extends StatelessWidget {
  const FirstScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('First Screen'),
      ),
      body: Center(
        child: Column(mainAxisSize: MainAxisSize.min, children: [
          Text(Navigator.of(context).canPop() ? 'Can Pop' : 'Can NOT Pop'),
          ElevatedButton(
            child: const Text('Second Page'),
            onPressed: () {
              Navigator.of(context).push(
                MaterialPageRoute(builder: (context) {
                  return const SecondScreen();
                }),
              );
            },
          ),
        ]),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  const SecondScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Second Screen'),
      ),
      body: Center(
        child: Column(mainAxisSize: MainAxisSize.min, children: [
          Container(
            width: 50,
            height: 50,
            color: Colors.yellow,
          ),
          const Text(
              'Change color of container for hot-reload before hitting the back button'),
          const BackButton(),
        ]),
      ),
    );
  }
}
`

**Comment by rashadNEON** [comment]
> > Has anyone found a reliable workaround for this?
> 
> Instead of `Navigator.canPop(context)` you can use `ModalRoute.of(context).canPop()`

Can you please explain why this works?

**Comment by exaby73** [comment]
I can no longer reproduce this issue on `Master (3.10.0-10.0.pre.17)`. Therefore, I will close this issue. If you can still reproduce this issue, please comment with steps to reproduce, a [complete and minimal, reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) and the output of `flutter doctor -v`. Thank you.

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.10.0-10.0.pre.17, on macOS 13.3.1 22E261 darwin-arm64, locale en-US)
    • Flutter version 3.10.0-10.0.pre.17 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.19.5/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b04efe4ecf (4 hours ago), 2023-04-20 01:31:24 -0400
    • Engine revision 2db85cbe6a
    • Dart version 3.1.0 (build 3.1.0-22.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.2
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 112.1.50.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>
